{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "caba62b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "from copy import copy\n",
    "from scipy.stats import norm\n",
    "from scipy import sparse\n",
    "from scipy import stats\n",
    "from sklearn.neighbors import KernelDensity\n",
    "from ete3 import Tree\n",
    "from typing import Tuple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2a8c9cb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.verbosity = 0\n",
    "sc.settings.figdir = '/Genomics/chanlab/blaw/TLS/sandbox/scRNA/TLS_integrated/'\n",
    "\n",
    "clusterColorsFile = \"/Genomics/chanlab/mchan/Adriano/TLS/TLS_TLSCL/20211102_clusterColorsTLSCL.p\"\n",
    "with open(clusterColorsFile,'rb') as fp:\n",
    "    colorDict = pickle.load(fp)\n",
    "    \n",
    "\n",
    "TLS_barcodes = ['Bar1', 'Bar2', 'Bar4', 'Bar5', 'Bar7', 'Bar8', 'Bar10', 'Bar11', 'Bar13', 'Bar16', 'Bar19', 'Bar22']\n",
    "TLSCL_barcodes = ['Bar3', 'Bar6', 'Bar9', 'Bar12', 'Bar14', 'Bar15', 'Bar20', 'Bar21', 'Bar23', 'Bar24']\n",
    "\n",
    "# Read in tables of the cell states for each cell\n",
    "TLS1_cell_state_table = pd.read_csv('/Genomics/chanlab/blaw/TLS/LineageTracer/scRNAseq/TLS_120h_1_cellBC_cellState.tsv', sep='\\t')\n",
    "TLS2_cell_state_table = pd.read_csv('/Genomics/chanlab/blaw/TLS/LineageTracer/scRNAseq/TLS_120h_2_cellBC_cellState.tsv', sep='\\t')\n",
    "TLS_M_cell_state_table = pd.read_csv('/Genomics/chanlab/blaw/TLS/LineageTracer/scRNAseq/TLS_TLSCL_cellBC_cellState.tsv', sep='\\t')\n",
    "TLS1_cell_state_table.set_index('cellBC', inplace = True)\n",
    "TLS2_cell_state_table.set_index('cellBC', inplace = True)\n",
    "TLS_M_cell_state_table.set_index('cellBC', inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b5fe89f",
   "metadata": {},
   "source": [
    "# Load in the data and add metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "637406d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the integrated object\n",
    "TLS_integrated = sc.read_h5ad('/Genomics/chanlab/blaw/TLS/raw_data/scRNA/Integrated/TLS_TLSCL_1_2_Time_integrated.h5ad')\n",
    "\n",
    "# Load the multiseq barcodes metadata\n",
    "multiseq_barcodes = pd.read_csv('/Genomics/chanlab/blaw/TLS/metadata/multiseq_barcodes.txt', sep='\\t')\n",
    "\n",
    "# Add the multiseq group numbers and whether the sample is a TLS or TLSCL (in the multiseq experiment)\n",
    "# The TLS replicates from TLS1, TLS2, TIME ones are counted as NA in both of these categories to seperate them from the multiseq\n",
    "multiseq_barcode_dict = dict(zip(multiseq_barcodes['Sequence'], multiseq_barcodes['ID']))\n",
    "TLS_TLSCL_dict = {}\n",
    "for key, value in multiseq_barcode_dict.items():\n",
    "    multiseq_barcode_dict[key] = str(int(value[3:]))\n",
    "for row in multiseq_barcodes.iterrows():\n",
    "    TLS_TLSCL_dict[row[1][2]] = 'TLS'\n",
    "for row in multiseq_barcodes[multiseq_barcodes['TLS ID'].str.startswith('TLSCL')].iterrows():\n",
    "    TLS_TLSCL_dict[row[1][2]] = 'TLSCL'\n",
    "    \n",
    "multiseq_barcode_dict['Doublet'] = 'Doublet'\n",
    "multiseq_barcode_dict['Negative'] = 'Negative'\n",
    "TLS_TLSCL_dict['Doublet'] = 'Doublet'\n",
    "TLS_TLSCL_dict['Negative'] = 'Negative'\n",
    "\n",
    "# Merge the multiseq group metadata with the entire integrated dataset\n",
    "TLS_integrated.obs['MultiSeqGroup'] = TLS_integrated.obs['MultiSeqBCseq'].map(multiseq_barcode_dict)\n",
    "TLS_integrated.obs['TLSCL'] = TLS_integrated.obs['MultiSeqBCseq'].map(TLS_TLSCL_dict)\n",
    "\n",
    "# Load the monocle position values\n",
    "pseudotime = pd.read_csv('/Genomics/chanlab/blaw/TLS/raw_data/monocle_results/TLS_TLSCL_1_2_Time_integrated_Monocle_Pos.txt', sep = '\\t', index_col = 0)\n",
    "\n",
    "# Rename the index column\n",
    "pseudotime.rename(columns = {'x': 'pseudotime'}, inplace = True)\n",
    "\n",
    "# Merge the pseudotime values into the integrated object\n",
    "TLS_integrated.obs = pd.merge(TLS_integrated.obs, pseudotime, left_index = True, right_index = True)\n",
    "\n",
    "# Recenter the pseudotime value to put the NMP median in the center since both trajectories originate from NMPs\n",
    "NMP_median = np.median(TLS_integrated.obs[TLS_integrated.obs['cell_state'] == 'NMPs']['pseudotime'].values)\n",
    "TLS_integrated.obs['NT-Somite_Traj'] = TLS_integrated.obs['pseudotime'] - NMP_median"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "619e7d99",
   "metadata": {},
   "source": [
    "# Identify NMPs that are in the lineage datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cbe0a3f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open TLS1 and TLS2 trees\n",
    "TLS1_loc = '/Genomics/chanlab/blaw/TLS/data/AM-DNA-097/lineage/2_lineage_reconstruction/AM-DNA-097_hybrid_newick_noMutationlessEdges_Labeled.nwk'\n",
    "t1 = Tree(TLS1_loc, format=1)\n",
    "\n",
    "TLS2_loc = '/Genomics/chanlab/blaw/TLS/data/AM-DNA-098/lineage/2_lineage_reconstruction/AM-DNA-098_hybrid_newick_noMutationlessEdges_Labeled.nwk'\n",
    "t2 = Tree(TLS2_loc, format=1)\n",
    "\n",
    "# Save lists of all the leaves that appear TLS1 and TLS2\n",
    "TLS1_leaves = [leaf.name for leaf in t1.get_leaves()]\n",
    "TLS2_leaves = [leaf.name for leaf in t2.get_leaves()]\n",
    "\n",
    "# Open TLS M and TLSCL trees to grab the leaves\n",
    "TLS_m_leaves = []\n",
    "TLSCL_leaves = []\n",
    "\n",
    "# For each barcode in TLS and TLSCL, open the tree and add the leaves (cells) to the total tree leaf lists\n",
    "for barcode in TLS_barcodes:\n",
    "    treeFile = \"/Genomics/chanlab/blaw/TLS/data/AM-DNA-258/lineage/3_lineage_reconstruction/{}/{}_newick_noMutationlessEdges_Labeled.nwk\".format(barcode, barcode)\n",
    "    t = Tree(treeFile, format = 1)\n",
    "\n",
    "    for leaf in t.get_leaves():\n",
    "        TLS_m_leaves.append(leaf.name)\n",
    "\n",
    "for barcode in TLSCL_barcodes:\n",
    "    treeFile = \"/Genomics/chanlab/blaw/TLS/data/AM-DNA-258/lineage/3_lineage_reconstruction/{}/{}_newick_noMutationlessEdges_Labeled.nwk\".format(barcode, barcode)\n",
    "    t = Tree(treeFile, format = 1)\n",
    "\n",
    "    for leaf in t.get_leaves():\n",
    "        TLSCL_leaves.append(leaf.name)\n",
    "        \n",
    "# Reformat the leaf names so that they can be used to index the integrated object\n",
    "tree_cells = ['TLS_120h_rep1_' + i for i in TLS1_leaves] + ['TLS_120h_rep2_' + i for i in TLS2_leaves] + ['TLS_TLSCL_' + i for i in TLS_m_leaves] + ['TLS_TLSCL_' + i for i in TLSCL_leaves]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "77c3ac44",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n"
     ]
    }
   ],
   "source": [
    "# Subset just the NMPs from the full object\n",
    "NMPs = TLS_integrated[TLS_integrated.obs['cell_state'] == 'NMPs'].copy()\n",
    "\n",
    "# Make a list of all the NMPs in the integrated object\n",
    "NMP_index = list(NMPs.obs.index)\n",
    "\n",
    "# Make a list of all the NMPs that appear on a tree\n",
    "NMP_tree_cells = []\n",
    "\n",
    "for i in tree_cells:\n",
    "    #print(i)\n",
    "    if i in NMP_index:\n",
    "        NMP_tree_cells.append(i)\n",
    "\n",
    "# copy a dataframe of the obs table of the NMPs that are in trees\n",
    "tree_NMPs = NMPs[NMPs.obs.index.isin(NMP_tree_cells)].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "181816df",
   "metadata": {},
   "source": [
    "# Label the NMPs by their neighboring cells on the tree\n",
    "\n",
    "- 'Drop' NMPs that are neighbors with a non-NMP cell type ('PGC', 'Endoderm', 'Endothelial'. Unassigned is +/-)\n",
    "- Proliferating <- NMPs that are neighbors with only other NMPs (or nothing)\n",
    "- Bipotent <- NMPs that are neighbors with both Neural and Somitic cells\n",
    "- Somitic <- NMPs that are neighbors with only somitic cells\n",
    "- Neural <- NMPs that are neighbors with only neural cells\n",
    "\n",
    "Also capture the fraction of somitic and fraction of neural neighbors for us to score the NMP 'commitment'\n",
    "- Commitment is scored as the # of somitic cells + 1 / # of neural cells + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7a8546f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add a column in the tree_NMPs table to classify the NMPs\n",
    "tree_NMPs.obs['NMP Type'] = ''\n",
    "tree_NMPs.obs['frac_Somitic'] = 0\n",
    "tree_NMPs.obs['frac_Neural'] = 0\n",
    "tree_NMPs.obs['Somite_Neural_ratio'] = 0\n",
    "\n",
    "# Populate the new column by manually looking at all nodes in TLS1\n",
    "for node in t1.traverse():\n",
    "    # find nodes so that we know the parent node for a leaf\n",
    "    if not node.is_leaf():\n",
    "        # check if any of the children are an NMP\n",
    "        for child in node.children:\n",
    "            if child.is_leaf() and TLS1_cell_state_table.loc[child.name, 'cell_state'] == 'NMPs':\n",
    "                # classify the NMP based on all the cell types that get made from the same parent\n",
    "                #leaves = [leaf.name for leaf in node.get_leaves()]\n",
    "                \n",
    "                # Classify the NMP based on the cell types of sibling cells (leafs that are children of the same parent node)\n",
    "                leaves = []\n",
    "                for subnode in node.children:\n",
    "                    if subnode.is_leaf():\n",
    "                        leaves.append(subnode.name)\n",
    "                \n",
    "                states = set()\n",
    "                neural_count = 0\n",
    "                somitic_count = 0\n",
    "                for leaf in leaves:\n",
    "                    if TLS1_cell_state_table.loc[leaf, 'cell_state'] in ['pPSM', 'aPSM', 'Somite', 'Somite0', 'Somite1', 'Somite-1', 'SomiteSclero', 'SomiteDermo']:\n",
    "                        states.add('Somitic')\n",
    "                        somitic_count += 1\n",
    "                    elif TLS1_cell_state_table.loc[leaf, 'cell_state'] in ['NeuralTube1', 'NeuralTube2']:\n",
    "                        states.add('Neural')\n",
    "                        neural_count += 1\n",
    "                    elif TLS1_cell_state_table.loc[leaf, 'cell_state'] in ['NMPs']:\n",
    "                        states.add('NMPs')\n",
    "                    elif TLS1_cell_state_table.loc[leaf, 'cell_state'] in ['Endothelial', 'Endoderm', 'PCGLC']:\n",
    "                        states.add('Drop')\n",
    "                        \n",
    "                tree_NMPs.obs.loc['TLS_120h_rep1_' + child.name, 'frac_Somitic'] = somitic_count / len(leaves)\n",
    "                tree_NMPs.obs.loc['TLS_120h_rep1_' + child.name, 'frac_Neural'] = neural_count / len(leaves)\n",
    "                tree_NMPs.obs.loc['TLS_120h_rep1_' + child.name, 'Somite_Neural_ratio'] = (somitic_count + 1) / (neural_count + 1)\n",
    "                        \n",
    "                if states == set(['Neural', 'NMPs']):\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep1_' + child.name, 'NMP Type'] = 'NMPs-Neural'\n",
    "                elif states == set(['Somitic', 'NMPs']):\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep1_' + child.name, 'NMP Type'] = 'NMPs-Somitic'\n",
    "                elif states == set(['Neural', 'Somitic', 'NMPs']):\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep1_' + child.name, 'NMP Type'] = 'NMPs-Bipotent'\n",
    "                elif states == set(['NMPs']):\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep1_' + child.name, 'NMP Type'] = 'NMPs-Renewing'\n",
    "                else:\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep1_' + child.name, 'NMP Type'] = 'Dropped'\n",
    "                    \n",
    "                    \n",
    "for node in t2.traverse():\n",
    "    # find nodes so that we know the parent node for a leaf\n",
    "    if not node.is_leaf():\n",
    "        # check if any of the children are an NMP\n",
    "        for child in node.children:\n",
    "            if child.is_leaf() and TLS2_cell_state_table.loc[child.name, 'cell_state'] == 'NMPs':\n",
    "                # classify the NMP based on all the cell types that get made from the same parent\n",
    "                #leaves = [leaf.name for leaf in node.get_leaves()]\n",
    "                \n",
    "                # Classify the NMP based on the cell types of sibling cells (leafs that are children of the same parent node)\n",
    "                leaves = []\n",
    "                for subnode in node.children:\n",
    "                    if subnode.is_leaf():\n",
    "                        leaves.append(subnode.name)\n",
    "                \n",
    "                states = set()\n",
    "                neural_count = 0\n",
    "                somitic_count = 0\n",
    "                for leaf in leaves:\n",
    "                    if TLS2_cell_state_table.loc[leaf, 'cell_state'] in ['pPSM', 'aPSM', 'Somite', 'Somite0', 'Somite1', 'Somite-1', 'SomiteSclero', 'SomiteDermo']:\n",
    "                        states.add('Somitic')\n",
    "                        somitic_count += 1\n",
    "                    elif TLS2_cell_state_table.loc[leaf, 'cell_state'] in ['NeuralTube1', 'NeuralTube2']:\n",
    "                        states.add('Neural')\n",
    "                        neural_count += 1\n",
    "                    elif TLS2_cell_state_table.loc[leaf, 'cell_state'] in ['NMPs']:\n",
    "                        states.add('NMPs')\n",
    "                    elif TLS2_cell_state_table.loc[leaf, 'cell_state'] in ['Endothelial', 'Endoderm', 'PCGLC']:\n",
    "                        states.add('Drop')\n",
    "                        \n",
    "                tree_NMPs.obs.loc['TLS_120h_rep2_' + child.name, 'frac_Somitic'] = somitic_count / len(leaves)\n",
    "                tree_NMPs.obs.loc['TLS_120h_rep2_' + child.name, 'frac_Neural'] = neural_count / len(leaves)\n",
    "                tree_NMPs.obs.loc['TLS_120h_rep2_' + child.name, 'Somite_Neural_ratio'] = (somitic_count + 1) / (neural_count + 1)\n",
    "                        \n",
    "                if states == set(['Neural', 'NMPs']):\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep2_' + child.name, 'NMP Type'] = 'NMPs-Neural'\n",
    "                elif states == set(['Somitic', 'NMPs']):\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep2_' + child.name, 'NMP Type'] = 'NMPs-Somitic'\n",
    "                elif states == set(['Neural', 'Somitic', 'NMPs']):\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep2_' + child.name, 'NMP Type'] = 'NMPs-Bipotent'\n",
    "                elif states == set(['NMPs']):\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep2_' + child.name, 'NMP Type'] = 'NMPs-Renewing'\n",
    "                else:\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep2_' + child.name, 'NMP Type'] = 'Dropped'\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "202ca082",
   "metadata": {},
   "outputs": [],
   "source": [
    "for barcode in TLS_barcodes + TLSCL_barcodes:\n",
    "    treeFile = \"/Genomics/chanlab/blaw/TLS/data/AM-DNA-258/lineage/3_lineage_reconstruction/{}/{}_newick_noMutationlessEdges_Labeled.nwk\".format(barcode, barcode)\n",
    "    t = Tree(treeFile, format = 1)\n",
    "    \n",
    "    for node in t.traverse():\n",
    "        # only do analysis on non leaves\n",
    "        if not node.is_leaf():\n",
    "            # check if any of the children are an NMP\n",
    "            for child in node.children:\n",
    "                if child.is_leaf() and TLS_M_cell_state_table.loc[child.name, 'cell_state'] == 'NMPs':\n",
    "                    #leaves = [leaf.name for leaf in node.get_leaves()]\n",
    "\n",
    "                    # Classify the NMP based on the cell types of sibling cells (leafs that are children of the same parent node)\n",
    "                    leaves = []\n",
    "                    for subnode in node.children:\n",
    "                        if subnode.is_leaf():\n",
    "                            leaves.append(subnode.name)\n",
    "                            \n",
    "                    states = set()\n",
    "                    neural_count = 0\n",
    "                    somitic_count = 0\n",
    "                    for leaf in leaves:\n",
    "                        if TLS_M_cell_state_table.loc[leaf, 'cell_state'] in ['pPSM', 'aPSM', 'Somite', 'Somite0', 'Somite1', 'Somite-1', 'SomiteSclero', 'SomiteDermo']:\n",
    "                            states.add('Somitic')\n",
    "                            somitic_count += 1\n",
    "                        elif TLS_M_cell_state_table.loc[leaf, 'cell_state'] in ['NeuralTube1', 'NeuralTube2']:\n",
    "                            states.add('Neural')\n",
    "                            neural_count += 1\n",
    "                        elif TLS_M_cell_state_table.loc[leaf, 'cell_state'] in ['NMPs']:\n",
    "                            states.add('NMPs')\n",
    "                        elif TLS_M_cell_state_table.loc[leaf, 'cell_state'] in ['Endoderm', 'Endothelial', 'PCGLC']:\n",
    "                            states.add('Drop')\n",
    "                            \n",
    "                    tree_NMPs.obs.loc['TLS_TLSCL_' + child.name, 'frac_Somitic'] = somitic_count / len(leaves)\n",
    "                    tree_NMPs.obs.loc['TLS_TLSCL_' + child.name, 'frac_Neural'] = neural_count / len(leaves)\n",
    "                    tree_NMPs.obs.loc['TLS_TLSCL_' + child.name, 'Somite_Neural_ratio'] = (somitic_count + 1) / (neural_count + 1)\n",
    "                        \n",
    "\n",
    "                    if states == set(['Neural', 'NMPs']):\n",
    "                        tree_NMPs.obs.loc['TLS_TLSCL_' + child.name, 'NMP Type'] = 'NMPs-Neural'\n",
    "                    elif states == set(['Somitic', 'NMPs']):\n",
    "                        tree_NMPs.obs.loc['TLS_TLSCL_' + child.name, 'NMP Type'] = 'NMPs-Somitic'\n",
    "                    elif states == set(['Neural', 'Somitic', 'NMPs']):\n",
    "                        tree_NMPs.obs.loc['TLS_TLSCL_' + child.name, 'NMP Type'] = 'NMPs-Bipotent'\n",
    "                    elif states == set(['NMPs']):\n",
    "                        tree_NMPs.obs.loc['TLS_TLSCL_' + child.name, 'NMP Type'] = 'NMPs-Renewing'\n",
    "                    else:\n",
    "                        tree_NMPs.obs.loc['TLS_TLSCL_' + child.name, 'NMP Type'] = 'Dropped'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b2798c88",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TLS_120h_rep1\n",
      "\t NMPs-Renewing 29\n",
      "\t NMPs-Neural 13\n",
      "\t NMPs-Somitic 24\n",
      "\t NMPs-Bipotent 24\n",
      "\t Dropped 32\n",
      "TLS_120h_rep2\n",
      "\t NMPs-Renewing 1\n",
      "\t NMPs-Neural 6\n",
      "\t NMPs-Somitic 3\n",
      "\t NMPs-Bipotent 0\n",
      "\t Dropped 3\n",
      "TLS_TLSCL\n",
      "\t NMPs-Renewing 1032\n",
      "\t NMPs-Neural 299\n",
      "\t NMPs-Somitic 330\n",
      "\t NMPs-Bipotent 52\n",
      "\t Dropped 7\n"
     ]
    }
   ],
   "source": [
    "for j in ['TLS_120h_rep1', 'TLS_120h_rep2', 'TLS_TLSCL']:\n",
    "    print(j)\n",
    "    for i in tree_NMPs.obs['NMP Type'].unique():\n",
    "        print('\\t', i, len(tree_NMPs.obs[(tree_NMPs.obs['orig.ident'] == j) & (tree_NMPs.obs['NMP Type'] == i)]['NMP Type']))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c3f4b3b",
   "metadata": {},
   "source": [
    "# DE between NMP types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7a8b9472",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_names(index, adata):\n",
    "    return adata.var.loc[index][0]\n",
    "\n",
    "def add_avg_expr(index, scipy_sparse):\n",
    "    return scipy_sparse[:, int(index)].mean()\n",
    "\n",
    "def add_frac_expressed(index, scipy_sparse):\n",
    "    cell_count = scipy_sparse.shape[0]\n",
    "    cells_expressing = scipy_sparse[:, int(index)].count_nonzero()\n",
    "    \n",
    "    return cells_expressing / cell_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4d9290d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculateDE(adata, column, comparisons, output_loc, sampleName):\n",
    "    '''\n",
    "    input:\n",
    "        adata - a scanpy object to run the analysis on\n",
    "        column - a column in the adata.obs table that will be used to separate the data into 2 categories\n",
    "        comparisons - a tuple that contains the 2 categories from the column to separate the data. The second is the ref name\n",
    "        output_loc - a string that points to a valid folder to save the files\n",
    "        name - a name for the \n",
    "    output:\n",
    "        a volcano plot of the resulting gene expression data (logfc vs -log10pval)\n",
    "        a plot of the scanpy outputs from the differential expression (top genes by score)\n",
    "        a table that stores the differential expression values for each gene\n",
    "    '''\n",
    "    # set the scanpy figure directory to output_loc\n",
    "    sc.settings.figdir = output_loc + 'plots/'\n",
    "    sc.settings.verbosity = 0\n",
    "    \n",
    "    upName = comparisons[0]\n",
    "    downName = comparisons[1]\n",
    "\n",
    "    # exclude the TLS1 and timecourse data, so i'm only using the TLS and TLSCL from the multiseq\n",
    "    temp_adata = adata[(adata.obs[column].isin(comparisons))].copy()\n",
    "    \n",
    "    temp_adata = temp_adata.raw.to_adata()\n",
    "    sc.pp.normalize_per_cell(temp_adata, counts_per_cell_after=1e4)\n",
    "    sc.pp.log1p(temp_adata)\n",
    "    \n",
    "    sc.tl.rank_genes_groups(temp_adata, column, method='wilcoxon', groups = [upName], reference = downName, key_added = \"wilcoxon\", gene_symbols = '_index')\n",
    "    sc.pl.rank_genes_groups(temp_adata, n_genes=25, sharey=False, key=\"wilcoxon\", gene_symbols = '_index', save = '_{}_vs_{}_{}'.format(upName, downName, sampleName), show = False)\n",
    "    \n",
    "    # record the information from the logfc table\n",
    "    names = [i[0] for i in temp_adata.uns['wilcoxon']['names'].tolist()]\n",
    "    scores = [i[0] for i in temp_adata.uns['wilcoxon']['scores'].tolist()]\n",
    "    pvals = [i[0] for i in temp_adata.uns['wilcoxon']['pvals'].tolist()]\n",
    "    pvals_adj = [i[0] for i in temp_adata.uns['wilcoxon']['pvals_adj'].tolist()]\n",
    "    logfcs = [i[0] for i in temp_adata.uns['wilcoxon']['logfoldchanges'].tolist()]\n",
    "    \n",
    "    # save it to a df\n",
    "    temp_df = pd.DataFrame(zip(names, scores, pvals, pvals_adj, logfcs), columns = ['name_index', 'scores', 'pvals', 'pvals_adj', 'logfoldchanges'])\n",
    "    temp_df['names'] = temp_df['name_index'].apply(lambda x: add_names(x, temp_adata))\n",
    "\n",
    "    # record the avg and frac expr for TLS and TLSCL for each gene\n",
    "    upName_scipy_sparse = sparse.csr_matrix(temp_adata[temp_adata.obs[column] == upName].X)\n",
    "    temp_df[upName + '_avg_expr'] = temp_df['name_index'].apply(lambda x: add_avg_expr(x, upName_scipy_sparse))\n",
    "    temp_df[upName + '_frac_expr'] = temp_df['name_index'].apply(lambda x: add_frac_expressed(x, upName_scipy_sparse))\n",
    "\n",
    "    downName_scipy_sparse = sparse.csr_matrix(temp_adata[temp_adata.obs[column] == downName].X)\n",
    "    temp_df[downName + '_avg_expr'] = temp_df['name_index'].apply(lambda x: add_avg_expr(x, downName_scipy_sparse))\n",
    "    temp_df[downName + '_frac_expr'] = temp_df['name_index'].apply(lambda x: add_frac_expressed(x, downName_scipy_sparse))\n",
    "    \n",
    "    temp_df['frac_diff'] = abs(temp_df[upName + '_frac_expr'] - temp_df[downName + '_frac_expr'])\n",
    "    temp_df['-log10pval'] = -np.log10(temp_df['pvals'])\n",
    "    temp_df['manual_LogFC'] = np.log2(temp_df[upName + '_avg_expr'] + 0.01) - np.log2(temp_df[downName + '_avg_expr'] + 0.01) \n",
    "    \n",
    "    # save the df\n",
    "    temp_df.to_csv(output_loc + 'gene_lists/{}_{}_{}_wilcox_DE_full.csv'.format(upName, downName, sampleName), index = False)\n",
    "    temp_df[(temp_df['logfoldchanges'] > 1) & (temp_df['-log10pval'] > 1.2)].to_csv(output_loc + 'gene_lists/{}_{}_{}_wilcox_DE_upregulated.csv'.format(upName, downName, sampleName), index = False)\n",
    "    temp_df[(temp_df['logfoldchanges'] < -1) & (temp_df['-log10pval'] > 1.2)].to_csv(output_loc + 'gene_lists/{}_{}_{}_wilcox_DE_downregulated.csv'.format(upName, downName, sampleName), index = False)\n",
    "    \n",
    "    # Plot the volcano plot of the differential expression\n",
    "    x = temp_df[(temp_df[upName + '_avg_expr'] != 0) & (temp_df[downName + '_avg_expr'] != 0)]['logfoldchanges'].values\n",
    "    y = -np.log10(temp_df[(temp_df[upName + '_avg_expr'] != 0) & (temp_df[downName + '_avg_expr'] != 0)]['pvals']).values\n",
    "    plt.plot(x, y, '.', alpha = 0.3, color = 'lightgray')\n",
    "\n",
    "    # highlight genes that are higher than 1 logfoldchange and higher than 1.2 -log10pval\n",
    "    x2 = []\n",
    "    y2 = []\n",
    "    count = 0\n",
    "    for i in x:\n",
    "        try:\n",
    "            if x[count] > 1 and y[count] > 1.2:\n",
    "                x2.append(x[count])\n",
    "                y2.append(y[count])\n",
    "            if x[count] < -1 and y[count] > 1.2:\n",
    "                x2.append(x[count])\n",
    "                y2.append(y[count])\n",
    "        except:\n",
    "            pass\n",
    "        count += 1\n",
    "    plt.plot(x2, y2, '.')\n",
    "    \n",
    "    # Plot Rspo3, Fgf8, and T\n",
    "    genes = ['Rspo3', 'Fgf8', 'T']\n",
    "    for gene in genes:\n",
    "        x3 = temp_df[temp_df['names'] == gene]['logfoldchanges']\n",
    "        y3 = -np.log10(temp_df[temp_df['names'] == gene]['pvals'])\n",
    "        plt.annotate(text = gene, xy = (x3, y3))\n",
    "\n",
    "    plt.xlabel('LogFC')\n",
    "    plt.ylabel('-log10(pval)')\n",
    "    plt.title('{} vs {} - {} Differential Expression'.format(upName, downName, sampleName))\n",
    "    plt.savefig(output_loc + 'plots/{}_vs_{}_{}_volcano.pdf'.format(upName, downName, sampleName), dpi = 300)\n",
    "    #plt.show()\n",
    "    plt.close()\n",
    "    \n",
    "    \n",
    "    \n",
    "    # Plot the volcano plot of the differential expression using manual LogFC\n",
    "    x = temp_df[(temp_df[upName + '_avg_expr'] != 0) & (temp_df[downName + '_avg_expr'] != 0)]['manual_LogFC'].values\n",
    "    y = -np.log10(temp_df[(temp_df[upName + '_avg_expr'] != 0) & (temp_df[downName + '_avg_expr'] != 0)]['pvals']).values\n",
    "    plt.plot(x, y, '.', alpha = 0.3, color = 'lightgray')\n",
    "\n",
    "    # highlight genes that are higher than 1 logfoldchange and higher than 1.2 -log10pval\n",
    "    x2 = []\n",
    "    y2 = []\n",
    "    count = 0\n",
    "    for i in x:\n",
    "        try:\n",
    "            if x[count] > 1 and y[count] > 1.2:\n",
    "                x2.append(x[count])\n",
    "                y2.append(y[count])\n",
    "            if x[count] < -1 and y[count] > 1.2:\n",
    "                x2.append(x[count])\n",
    "                y2.append(y[count])\n",
    "        except:\n",
    "            pass\n",
    "        count += 1\n",
    "    plt.plot(x2, y2, '.')\n",
    "    \n",
    "    # Plot Rspo3, Fgf8, and T\n",
    "    genes = ['Rspo3', 'Fgf8', 'T']\n",
    "    for gene in genes:\n",
    "        x3 = temp_df[temp_df['names'] == gene]['manual_LogFC']\n",
    "        y3 = -np.log10(temp_df[temp_df['names'] == gene]['pvals'])\n",
    "        plt.annotate(text = gene, xy = (x3, y3))\n",
    "\n",
    "    plt.xlabel('LogFC')\n",
    "    plt.ylabel('-log10(pval)')\n",
    "    plt.title('{} vs {} - {} Differential Expression - Manual LogFC'.format(upName, downName, sampleName))\n",
    "    plt.savefig(output_loc + 'plots/{}_vs_{}_{}_volcano_man_LogFC.pdf'.format(upName, downName, sampleName), dpi = 300)\n",
    "    #plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7e22aca1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n"
     ]
    }
   ],
   "source": [
    "# subset just TLS1 and TLS2 NMPs\n",
    "tree_NMPs_TLS12 = tree_NMPs[(tree_NMPs.obs['orig.ident'].isin(['TLS_120h_rep1', 'TLS_120h_rep2']))].copy()\n",
    "\n",
    "# Run DE between each of the NMP types in TLS1 and TLS2\n",
    "for i in [('NMPs-Somitic', 'NMPs-Neural'), ('NMPs-Bipotent', 'NMPs-Neural'), ('NMPs-Renewing', 'NMPs-Neural'), ('NMPs-Somitic', 'NMPs-Bipotent'), ('NMPs-Renewing', 'NMPs-Bipotent'), ('NMPs-Somitic', 'NMPs-Renewing')]:\n",
    "    calculateDE(tree_NMPs_TLS12,\n",
    "                'NMP Type',\n",
    "                i,\n",
    "                '/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/1_DE/TLS1_2_NMP_type/',\n",
    "                'TLS12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4be9d37a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n"
     ]
    }
   ],
   "source": [
    "# subset and run DE on TLS M\n",
    "tree_NMPs_M = tree_NMPs[(tree_NMPs.obs['orig.ident'] == 'TLS_TLSCL') & (tree_NMPs.obs['TLSCL'] == 'TLS')].copy()\n",
    "\n",
    "for i in [('NMPs-Somitic', 'NMPs-Neural'), ('NMPs-Bipotent', 'NMPs-Neural'), ('NMPs-Renewing', 'NMPs-Neural'), ('NMPs-Somitic', 'NMPs-Bipotent'), ('NMPs-Renewing', 'NMPs-Bipotent'), ('NMPs-Somitic', 'NMPs-Renewing')]:\n",
    "    calculateDE(tree_NMPs_M,\n",
    "                'NMP Type',\n",
    "                i,\n",
    "                '/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/1_DE/TLS1_2_NMP_type/',\n",
    "                'TLS_M')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "f24dfc6b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "... storing 'NMP Type' as categorical\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n"
     ]
    }
   ],
   "source": [
    "# Subset and run DE on TLSCL\n",
    "tree_NMPs_TLSCL = tree_NMPs[(tree_NMPs.obs['orig.ident'] == 'TLS_TLSCL') & (tree_NMPs.obs['TLSCL'] == 'TLSCL') & (tree_NMPs.obs['NMP Type'].isin(['NMPs-Renewing', 'NMPs-Neural', 'NMPs-Somitic', 'NMPs-Bipotent']))].copy()\n",
    "\n",
    "for i in [('NMPs-Somitic', 'NMPs-Neural'), ('NMPs-Bipotent', 'NMPs-Neural'), ('NMPs-Renewing', 'NMPs-Neural'), ('NMPs-Somitic', 'NMPs-Bipotent'), ('NMPs-Renewing', 'NMPs-Bipotent'), ('NMPs-Somitic', 'NMPs-Renewing')]:\n",
    "    calculateDE(tree_NMPs_TLSCL,\n",
    "                'NMP Type',\n",
    "                i,\n",
    "                '/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/1_DE/TLS1_2_NMP_type/',\n",
    "                'TLSCL')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
