{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "929c7dbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "from copy import copy\n",
    "from matplotlib import rc_context\n",
    "from scipy.stats import norm\n",
    "from scipy import sparse\n",
    "from scipy import stats\n",
    "from sklearn.neighbors import KernelDensity\n",
    "from ete3 import Tree\n",
    "from typing import Tuple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2ff8a8ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.verbosity = 0\n",
    "sc.settings.figdir = '/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/2_pseudotime/'\n",
    "\n",
    "clusterColorsFile = \"/Genomics/chanlab/mchan/Adriano/TLS/TLS_TLSCL/20211102_clusterColorsTLSCL.p\"\n",
    "with open(clusterColorsFile,'rb') as fp:\n",
    "    colorDict = pickle.load(fp)\n",
    "    \n",
    "\n",
    "TLS_barcodes = ['Bar1', 'Bar2', 'Bar4', 'Bar5', 'Bar7', 'Bar8', 'Bar10', 'Bar11', 'Bar13', 'Bar16', 'Bar19', 'Bar22']\n",
    "TLSCL_barcodes = ['Bar3', 'Bar6', 'Bar9', 'Bar12', 'Bar14', 'Bar15', 'Bar20', 'Bar21', 'Bar23', 'Bar24']\n",
    "\n",
    "# Read in tables of the cell states for each cell\n",
    "TLS1_cell_state_table = pd.read_csv('/Genomics/chanlab/blaw/TLS/LineageTracer/scRNAseq/TLS_120h_1_cellBC_cellState.tsv', sep='\\t')\n",
    "TLS2_cell_state_table = pd.read_csv('/Genomics/chanlab/blaw/TLS/LineageTracer/scRNAseq/TLS_120h_2_cellBC_cellState.tsv', sep='\\t')\n",
    "TLS_M_cell_state_table = pd.read_csv('/Genomics/chanlab/blaw/TLS/LineageTracer/scRNAseq/TLS_TLSCL_cellBC_cellState.tsv', sep='\\t')\n",
    "TLS1_cell_state_table.set_index('cellBC', inplace = True)\n",
    "TLS2_cell_state_table.set_index('cellBC', inplace = True)\n",
    "TLS_M_cell_state_table.set_index('cellBC', inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "63a396c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the integrated object\n",
    "TLS_integrated = sc.read_h5ad('/Genomics/chanlab/blaw/TLS/raw_data/scRNA/Integrated/TLS_TLSCL_1_2_Time_integrated.h5ad')\n",
    "\n",
    "# Load the multiseq barcodes metadata\n",
    "multiseq_barcodes = pd.read_csv('/Genomics/chanlab/blaw/TLS/metadata/multiseq_barcodes.txt', sep='\\t')\n",
    "\n",
    "# Add the multiseq group numbers and whether the sample is a TLS or TLSCL (in the multiseq experiment)\n",
    "# The TLS replicates from TLS1, TLS2, TIME ones are counted as NA in both of these categories to seperate them from the multiseq\n",
    "multiseq_barcode_dict = dict(zip(multiseq_barcodes['Sequence'], multiseq_barcodes['ID']))\n",
    "TLS_TLSCL_dict = {}\n",
    "for key, value in multiseq_barcode_dict.items():\n",
    "    multiseq_barcode_dict[key] = str(int(value[3:]))\n",
    "for row in multiseq_barcodes.iterrows():\n",
    "    TLS_TLSCL_dict[row[1][2]] = 'TLS'\n",
    "for row in multiseq_barcodes[multiseq_barcodes['TLS ID'].str.startswith('TLSCL')].iterrows():\n",
    "    TLS_TLSCL_dict[row[1][2]] = 'TLSCL'\n",
    "    \n",
    "multiseq_barcode_dict['Doublet'] = 'Doublet'\n",
    "multiseq_barcode_dict['Negative'] = 'Negative'\n",
    "TLS_TLSCL_dict['Doublet'] = 'Doublet'\n",
    "TLS_TLSCL_dict['Negative'] = 'Negative'\n",
    "\n",
    "# Merge the multiseq group metadata with the entire integrated dataset\n",
    "TLS_integrated.obs['MultiSeqGroup'] = TLS_integrated.obs['MultiSeqBCseq'].map(multiseq_barcode_dict)\n",
    "TLS_integrated.obs['TLSCL'] = TLS_integrated.obs['MultiSeqBCseq'].map(TLS_TLSCL_dict)\n",
    "\n",
    "# Load the monocle position values\n",
    "pseudotime = pd.read_csv('/Genomics/chanlab/blaw/TLS/raw_data/monocle_results/TLS_TLSCL_1_2_Time_integrated_Monocle_Pos.txt', sep = '\\t', index_col = 0)\n",
    "\n",
    "# Rename the index column\n",
    "pseudotime.rename(columns = {'x': 'pseudotime'}, inplace = True)\n",
    "\n",
    "# Merge the pseudotime values into the integrated object\n",
    "TLS_integrated.obs = pd.merge(TLS_integrated.obs, pseudotime, left_index = True, right_index = True)\n",
    "\n",
    "# Recenter the pseudotime value to put the NMP median in the center since both trajectories originate from NMPs\n",
    "NMP_median = np.median(TLS_integrated.obs[TLS_integrated.obs['cell_state'] == 'NMPs']['pseudotime'].values)\n",
    "TLS_integrated.obs['NT-Somite_Traj'] = TLS_integrated.obs['pseudotime'] - NMP_median"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9b60982",
   "metadata": {},
   "source": [
    "# look at pseudotime over the full TLS dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3d221162",
   "metadata": {},
   "outputs": [],
   "source": [
    "# restrict the analysis to just NMP cells that are part of the pseudotime axis\n",
    "NMP_state_list = ['NMPs', 'NeuralTube1', 'NeuralTube2', 'Somite', 'Somite0', 'Somite-1', 'SomiteDermo', 'SomiteSclero', 'aPSM', 'pPSM']\n",
    "\n",
    "NMP_states = TLS_integrated[(TLS_integrated.obs['cell_state'].isin(NMP_state_list)) & (TLS_integrated.obs['NT-Somite_Traj'] < 100)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0716b5fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'orig.ident' as categorical\n",
      "... storing 'integrated_snn_res.0.5' as categorical\n",
      "... storing 'seurat_clusters' as categorical\n",
      "... storing 'BC' as categorical\n",
      "... storing 'Phase' as categorical\n",
      "... storing 'old.ident' as categorical\n",
      "... storing 'integrated_snn_res.0.8' as categorical\n",
      "... storing 'cluster' as categorical\n",
      "... storing 'cell_state' as categorical\n",
      "... storing 'predicted.id' as categorical\n",
      "... storing 'cellBC' as categorical\n",
      "... storing 'MultiSeqBCseq' as categorical\n",
      "... storing 'MultiSeqGroup' as categorical\n",
      "... storing 'TLSCL' as categorical\n"
     ]
    }
   ],
   "source": [
    "with rc_context({'figure.figsize': (5, 5)}):\n",
    "    sc.pl.umap(NMP_states, color=['NT-Somite_Traj', 'cell_state'], palette = colorDict, title='NT - Somite Trajectory', save = 'NMP_states', show = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "62e8d2ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Collect the pseduotime values of each cell in the whole dataset\n",
    "all_pseudotime_values = NMP_states.obs['NT-Somite_Traj']\n",
    "\n",
    "# Collect the pseudotime values of each cell in the timecourse\n",
    "TLS_96_pseudotime_values = NMP_states.obs[NMP_states.obs['orig.ident'] == 'mGast_96h']['NT-Somite_Traj']\n",
    "TLS_108_pseudotime_values = NMP_states.obs[NMP_states.obs['orig.ident'] == 'mGast_108h']['NT-Somite_Traj']\n",
    "TLS_120_pseudotime_values = NMP_states.obs[NMP_states.obs['orig.ident'] == 'mGast_120h']['NT-Somite_Traj']\n",
    "\n",
    "# Collect the pseudotime values of each cell in TLS1 and TLS2\n",
    "TLS1_pseudotime_values = NMP_states.obs[NMP_states.obs['orig.ident'] == 'TLS_120h_rep1']['NT-Somite_Traj']\n",
    "TLS2_pseudotime_values = NMP_states.obs[NMP_states.obs['orig.ident'] == 'TLS_120h_rep2']['NT-Somite_Traj']\n",
    "\n",
    "# Collect the pseudotime values of each cell in TLS and TLSCL\n",
    "TLS_m_pseudotime_values = NMP_states.obs[(NMP_states.obs['TLSCL'] == 'TLS')]['NT-Somite_Traj']\n",
    "TLSCL_m_pseudotime_values = NMP_states.obs[(NMP_states.obs['TLSCL'] == 'TLSCL')]['NT-Somite_Traj']\n",
    "\n",
    "# Plot the density results\n",
    "fig, axs = plt.subplots(2, 2, figsize = (10, 5))\n",
    "\n",
    "ax1 = axs[0, 0]\n",
    "sns.kdeplot(all_pseudotime_values, color = 'black', ax = ax1)\n",
    "ax1.legend(['TLS_integrated'])\n",
    "ax1.set_xlim(-60, 30)\n",
    "ax1.set_ylim(0, 0.125)\n",
    "\n",
    "ax2 = axs[0, 1]\n",
    "sns.kdeplot(TLS_96_pseudotime_values, color = 'red', ax = ax2)\n",
    "sns.kdeplot(TLS_108_pseudotime_values, color = 'green', ax = ax2)\n",
    "sns.kdeplot(TLS_120_pseudotime_values, color = 'blue', ax = ax2)\n",
    "ax2.legend(['TLS_96h', 'TLS_108h', 'TLS_120h'])\n",
    "ax2.set_xlim(-60, 30)\n",
    "ax2.set_ylim(0, 0.125)\n",
    "\n",
    "\n",
    "ax3 = axs[1, 0]\n",
    "sns.kdeplot(TLS1_pseudotime_values, color = 'darkgreen', ax = ax3)\n",
    "sns.kdeplot(TLS2_pseudotime_values, color = 'pink', ax = ax3)\n",
    "ax3.legend(['TLS1', 'TLS2'])\n",
    "ax3.set_xlim(-60, 30)\n",
    "ax3.set_ylim(0, 0.125)\n",
    "\n",
    "ax4 = axs[1, 1]\n",
    "sns.kdeplot(TLS_m_pseudotime_values, color = 'darkgreen', ax = ax4)\n",
    "sns.kdeplot(TLSCL_m_pseudotime_values, color = 'pink', ax = ax4)\n",
    "ax4.legend(['TLS_Replicates', 'TLSCL_Replicates'])\n",
    "ax4.set_xlim(-60, 30)\n",
    "ax4.set_ylim(0, 0.125)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/2_pseudotime/pseudotime_density.pdf', dpi = 300)\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f7a00e0",
   "metadata": {},
   "source": [
    "# Look at NMPs over pseudotime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b49cf1d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "NMPs = TLS_integrated[TLS_integrated.obs['cell_state'] == 'NMPs'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bf788b7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# NMP specific density plots\n",
    "all_NMP_pseudotime_values = NMPs.obs['NT-Somite_Traj']\n",
    "TLS_96_NMP_pseudotime_values = NMPs.obs[(NMPs.obs['orig.ident'] == 'mGast_96h')]['NT-Somite_Traj']\n",
    "TLS_108_NMP_pseudotime_values = NMPs.obs[(NMPs.obs['orig.ident'] == 'mGast_108h')]['NT-Somite_Traj']\n",
    "TLS_120_NMP_pseudotime_values = NMPs.obs[(NMPs.obs['orig.ident'] == 'mGast_120h')]['NT-Somite_Traj']\n",
    "\n",
    "TLS1_NMP_pseudotime_values = NMPs.obs[(NMPs.obs['orig.ident'] == 'TLS_120h_rep1')]['NT-Somite_Traj']\n",
    "TLS2_NMP_pseudotime_values = NMPs.obs[(NMPs.obs['orig.ident'] == 'TLS_120h_rep2')]['NT-Somite_Traj']\n",
    "\n",
    "TLS_m_NMP_pseudotime_values = NMPs.obs[(NMPs.obs['TLSCL'] == 'TLS')]['NT-Somite_Traj']\n",
    "TLSCL_m_NMP_pseudotime_values = NMPs.obs[(NMPs.obs['TLSCL'] == 'TLSCL')]['NT-Somite_Traj']\n",
    "\n",
    "fig, axs = plt.subplots(2, 2, figsize = (10, 5))\n",
    "\n",
    "ax1 = axs[0, 0]\n",
    "sns.kdeplot(all_NMP_pseudotime_values, color = 'black', ax = ax1)\n",
    "ax1.legend(['TLS_integrated'])\n",
    "ax1.set_xlim(-20, 20)\n",
    "ax1.set_ylim(0, 0.5)\n",
    "\n",
    "ax2 = axs[0, 1]\n",
    "sns.kdeplot(TLS_96_NMP_pseudotime_values, color = 'red', ax = ax2)\n",
    "sns.kdeplot(TLS_108_NMP_pseudotime_values, color = 'green', ax = ax2)\n",
    "sns.kdeplot(TLS_120_NMP_pseudotime_values, color = 'blue', ax = ax2)\n",
    "ax2.legend(['TLS_96h', 'TLS_108h', 'TLS_120h'])\n",
    "ax2.set_xlim(-20, 20)\n",
    "ax2.set_ylim(0, 0.5)\n",
    "\n",
    "\n",
    "ax3 = axs[1, 0]\n",
    "sns.kdeplot(TLS1_NMP_pseudotime_values, color = 'darkgreen', ax = ax3)\n",
    "sns.kdeplot(TLS2_NMP_pseudotime_values, color = 'pink', ax = ax3)\n",
    "ax3.legend(['TLS1', 'TLS2'])\n",
    "ax3.set_xlim(-20, 20)\n",
    "ax3.set_ylim(0, 0.5)\n",
    "\n",
    "ax4 = axs[1, 1]\n",
    "sns.kdeplot(TLS_m_NMP_pseudotime_values, color = 'darkgreen', ax = ax4)\n",
    "sns.kdeplot(TLSCL_m_NMP_pseudotime_values, color = 'pink', ax = ax4)\n",
    "ax4.legend(['TLS_Replicates', 'TLSCL_Replicates'])\n",
    "ax4.set_xlim(-20, 20)\n",
    "ax4.set_ylim(0, 0.5)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/2_pseudotime/pseudotime_NMP_density.pdf', dpi = 300)\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd68f2d9",
   "metadata": {},
   "source": [
    "# Classify NMPs based off their lineage neighbors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b8da7170",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open TLS1 and TLS2 trees\n",
    "TLS1_loc = '/Genomics/chanlab/blaw/TLS/data/AM-DNA-097/lineage/2_lineage_reconstruction/AM-DNA-097_hybrid_newick_noMutationlessEdges_Labeled.nwk'\n",
    "t1 = Tree(TLS1_loc, format=1)\n",
    "\n",
    "TLS2_loc = '/Genomics/chanlab/blaw/TLS/data/AM-DNA-098/lineage/2_lineage_reconstruction/AM-DNA-098_hybrid_newick_noMutationlessEdges_Labeled.nwk'\n",
    "t2 = Tree(TLS2_loc, format=1)\n",
    "\n",
    "# Save lists of all the leaves that appear TLS1 and TLS2\n",
    "TLS1_leaves = [leaf.name for leaf in t1.get_leaves()]\n",
    "TLS2_leaves = [leaf.name for leaf in t2.get_leaves()]\n",
    "\n",
    "# Open TLS M and TLSCL trees to grab the leaves\n",
    "TLS_m_leaves = []\n",
    "TLSCL_leaves = []\n",
    "\n",
    "# For each barcode in TLS and TLSCL, open the tree and add the leaves (cells) to the total tree leaf lists\n",
    "for barcode in TLS_barcodes:\n",
    "    treeFile = \"/Genomics/chanlab/blaw/TLS/data/AM-DNA-258/lineage/3_lineage_reconstruction/{}/{}_newick_noMutationlessEdges_Labeled.nwk\".format(barcode, barcode)\n",
    "    t = Tree(treeFile, format = 1)\n",
    "\n",
    "    for leaf in t.get_leaves():\n",
    "        TLS_m_leaves.append(leaf.name)\n",
    "\n",
    "for barcode in TLSCL_barcodes:\n",
    "    treeFile = \"/Genomics/chanlab/blaw/TLS/data/AM-DNA-258/lineage/3_lineage_reconstruction/{}/{}_newick_noMutationlessEdges_Labeled.nwk\".format(barcode, barcode)\n",
    "    t = Tree(treeFile, format = 1)\n",
    "\n",
    "    for leaf in t.get_leaves():\n",
    "        TLSCL_leaves.append(leaf.name)\n",
    "        \n",
    "# Reformat the leaf names so that they can be used to index the integrated object\n",
    "tree_cells = ['TLS_120h_rep1_' + i for i in TLS1_leaves] + ['TLS_120h_rep2_' + i for i in TLS2_leaves] + ['TLS_TLSCL_' + i for i in TLS_m_leaves] + ['TLS_TLSCL_' + i for i in TLSCL_leaves]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7b173eee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a list of all the NMPs in the integrated object\n",
    "NMP_index = list(NMPs.obs.index)\n",
    "\n",
    "# Make a list of all the NMPs that appear on a tree\n",
    "NMP_tree_cells = []\n",
    "\n",
    "for i in tree_cells:\n",
    "    #print(i)\n",
    "    if i in NMP_index:\n",
    "        NMP_tree_cells.append(i)\n",
    "\n",
    "# copy a dataframe of the obs table of the NMPs that are in trees\n",
    "tree_NMPs = NMPs[NMPs.obs.index.isin(NMP_tree_cells)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8fcc6344",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check if lineage tracing causes NMPs to behave differently over the pseudotime\n",
    "tree_NMPs_values = list(tree_NMPs.obs['NT-Somite_Traj'])\n",
    "all_NMPs_values = list(NMPs.obs[NMPs.obs['orig.ident'].isin(['TLS_120h_rep1', 'TLS_120h_rep2', 'TLS_TLSCL'])]['NT-Somite_Traj'])\n",
    "\n",
    "fig, ax = plt.subplots(1, 1, figsize = (5, 2.5))\n",
    "\n",
    "sns.kdeplot(all_NMPs_values, color = 'black')\n",
    "sns.kdeplot(tree_NMPs_values, color = 'red')\n",
    "ax.legend(['All NMPs', 'Tree NMPs'])\n",
    "ax.set_title('Tree NMPs Distribution')\n",
    "ax.set_xlim(-20, 20)\n",
    "ax.set_ylim(0, 0.5)\n",
    "plt.savefig('/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/2_pseudotime/pseudotime_Tree_NMP_density.pdf', dpi = 300)\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cc43ee24",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add a column in the tree_NMPs table to classify the NMPs\n",
    "tree_NMPs.obs['NMP Type'] = ''\n",
    "tree_NMPs.obs['frac_Somitic'] = 0\n",
    "tree_NMPs.obs['frac_Neural'] = 0\n",
    "tree_NMPs.obs['Somite_Neural_ratio'] = 0\n",
    "\n",
    "# Populate the new column by manually looking at all nodes in TLS1\n",
    "for node in t1.traverse():\n",
    "    # find nodes so that we know the parent node for a leaf\n",
    "    if not node.is_leaf():\n",
    "        # check if any of the children are an NMP\n",
    "        for child in node.children:\n",
    "            if child.is_leaf() and TLS1_cell_state_table.loc[child.name, 'cell_state'] == 'NMPs':\n",
    "                # classify the NMP based on all the cell types that get made from the same parent\n",
    "                #leaves = [leaf.name for leaf in node.get_leaves()]\n",
    "                \n",
    "                # Classify the NMP based on the cell types of sibling cells (leafs that are children of the same parent node)\n",
    "                leaves = []\n",
    "                for subnode in node.children:\n",
    "                    if subnode.is_leaf():\n",
    "                        leaves.append(subnode.name)\n",
    "                \n",
    "                states = set()\n",
    "                neural_count = 0\n",
    "                somitic_count = 0\n",
    "                for leaf in leaves:\n",
    "                    if TLS1_cell_state_table.loc[leaf, 'cell_state'] in ['pPSM', 'aPSM', 'Somite', 'Somite0', 'Somite1', 'Somite-1', 'SomiteSclero', 'SomiteDermo']:\n",
    "                        states.add('Somitic')\n",
    "                        somitic_count += 1\n",
    "                    elif TLS1_cell_state_table.loc[leaf, 'cell_state'] in ['NeuralTube1', 'NeuralTube2']:\n",
    "                        states.add('Neural')\n",
    "                        neural_count += 1\n",
    "                    elif TLS1_cell_state_table.loc[leaf, 'cell_state'] in ['NMPs']:\n",
    "                        states.add('NMPs')\n",
    "                    elif TLS1_cell_state_table.loc[leaf, 'cell_state'] in ['Endothelial', 'Endoderm', 'PCGLC']:\n",
    "                        states.add('Drop')\n",
    "                        \n",
    "                tree_NMPs.obs.loc['TLS_120h_rep1_' + child.name, 'frac_Somitic'] = somitic_count / len(leaves)\n",
    "                tree_NMPs.obs.loc['TLS_120h_rep1_' + child.name, 'frac_Neural'] = neural_count / len(leaves)\n",
    "                tree_NMPs.obs.loc['TLS_120h_rep1_' + child.name, 'Somite_Neural_ratio'] = (somitic_count + 1) / (neural_count + 1)\n",
    "                        \n",
    "                if states == set(['Neural', 'NMPs']):\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep1_' + child.name, 'NMP Type'] = 'NMPs-Neural'\n",
    "                elif states == set(['Somitic', 'NMPs']):\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep1_' + child.name, 'NMP Type'] = 'NMPs-Somitic'\n",
    "                elif states == set(['Neural', 'Somitic', 'NMPs']):\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep1_' + child.name, 'NMP Type'] = 'NMPs-Bipotent'\n",
    "                elif states == set(['NMPs']):\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep1_' + child.name, 'NMP Type'] = 'NMPs-Renewing'\n",
    "                else:\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep1_' + child.name, 'NMP Type'] = 'Dropped'\n",
    "                    \n",
    "                    \n",
    "for node in t2.traverse():\n",
    "    # find nodes so that we know the parent node for a leaf\n",
    "    if not node.is_leaf():\n",
    "        # check if any of the children are an NMP\n",
    "        for child in node.children:\n",
    "            if child.is_leaf() and TLS2_cell_state_table.loc[child.name, 'cell_state'] == 'NMPs':\n",
    "                # classify the NMP based on all the cell types that get made from the same parent\n",
    "                #leaves = [leaf.name for leaf in node.get_leaves()]\n",
    "                \n",
    "                # Classify the NMP based on the cell types of sibling cells (leafs that are children of the same parent node)\n",
    "                leaves = []\n",
    "                for subnode in node.children:\n",
    "                    if subnode.is_leaf():\n",
    "                        leaves.append(subnode.name)\n",
    "                \n",
    "                states = set()\n",
    "                neural_count = 0\n",
    "                somitic_count = 0\n",
    "                for leaf in leaves:\n",
    "                    if TLS2_cell_state_table.loc[leaf, 'cell_state'] in ['pPSM', 'aPSM', 'Somite', 'Somite0', 'Somite1', 'Somite-1', 'SomiteSclero', 'SomiteDermo']:\n",
    "                        states.add('Somitic')\n",
    "                        somitic_count += 1\n",
    "                    elif TLS2_cell_state_table.loc[leaf, 'cell_state'] in ['NeuralTube1', 'NeuralTube2']:\n",
    "                        states.add('Neural')\n",
    "                        neural_count += 1\n",
    "                    elif TLS2_cell_state_table.loc[leaf, 'cell_state'] in ['NMPs']:\n",
    "                        states.add('NMPs')\n",
    "                    elif TLS2_cell_state_table.loc[leaf, 'cell_state'] in ['Endothelial', 'Endoderm', 'PCGLC']:\n",
    "                        states.add('Drop')\n",
    "                        \n",
    "                tree_NMPs.obs.loc['TLS_120h_rep2_' + child.name, 'frac_Somitic'] = somitic_count / len(leaves)\n",
    "                tree_NMPs.obs.loc['TLS_120h_rep2_' + child.name, 'frac_Neural'] = neural_count / len(leaves)\n",
    "                tree_NMPs.obs.loc['TLS_120h_rep2_' + child.name, 'Somite_Neural_ratio'] = (somitic_count + 1) / (neural_count + 1)\n",
    "                        \n",
    "                if states == set(['Neural', 'NMPs']):\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep2_' + child.name, 'NMP Type'] = 'NMPs-Neural'\n",
    "                elif states == set(['Somitic', 'NMPs']):\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep2_' + child.name, 'NMP Type'] = 'NMPs-Somitic'\n",
    "                elif states == set(['Neural', 'Somitic', 'NMPs']):\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep2_' + child.name, 'NMP Type'] = 'NMPs-Bipotent'\n",
    "                elif states == set(['NMPs']):\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep2_' + child.name, 'NMP Type'] = 'NMPs-Renewing'\n",
    "                else:\n",
    "                    tree_NMPs.obs.loc['TLS_120h_rep2_' + child.name, 'NMP Type'] = 'Dropped'\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c6fa6e44",
   "metadata": {},
   "outputs": [],
   "source": [
    "for barcode in TLS_barcodes + TLSCL_barcodes:\n",
    "    treeFile = \"/Genomics/chanlab/blaw/TLS/data/AM-DNA-258/lineage/3_lineage_reconstruction/{}/{}_newick_noMutationlessEdges_Labeled.nwk\".format(barcode, barcode)\n",
    "    t = Tree(treeFile, format = 1)\n",
    "    \n",
    "    for node in t.traverse():\n",
    "        # only do analysis on non leaves\n",
    "        if not node.is_leaf():\n",
    "            # check if any of the children are an NMP\n",
    "            for child in node.children:\n",
    "                if child.is_leaf() and TLS_M_cell_state_table.loc[child.name, 'cell_state'] == 'NMPs':\n",
    "                    #leaves = [leaf.name for leaf in node.get_leaves()]\n",
    "\n",
    "                    # Classify the NMP based on the cell types of sibling cells (leafs that are children of the same parent node)\n",
    "                    leaves = []\n",
    "                    for subnode in node.children:\n",
    "                        if subnode.is_leaf():\n",
    "                            leaves.append(subnode.name)\n",
    "                            \n",
    "                    states = set()\n",
    "                    neural_count = 0\n",
    "                    somitic_count = 0\n",
    "                    for leaf in leaves:\n",
    "                        if TLS_M_cell_state_table.loc[leaf, 'cell_state'] in ['pPSM', 'aPSM', 'Somite', 'Somite0', 'Somite1', 'Somite-1', 'SomiteSclero', 'SomiteDermo']:\n",
    "                            states.add('Somitic')\n",
    "                            somitic_count += 1\n",
    "                        elif TLS_M_cell_state_table.loc[leaf, 'cell_state'] in ['NeuralTube1', 'NeuralTube2']:\n",
    "                            states.add('Neural')\n",
    "                            neural_count += 1\n",
    "                        elif TLS_M_cell_state_table.loc[leaf, 'cell_state'] in ['NMPs']:\n",
    "                            states.add('NMPs')\n",
    "                        elif TLS_M_cell_state_table.loc[leaf, 'cell_state'] in ['Endoderm', 'Endothelial', 'PCGLC']:\n",
    "                            states.add('Drop')\n",
    "                            \n",
    "                    tree_NMPs.obs.loc['TLS_TLSCL_' + child.name, 'frac_Somitic'] = somitic_count / len(leaves)\n",
    "                    tree_NMPs.obs.loc['TLS_TLSCL_' + child.name, 'frac_Neural'] = neural_count / len(leaves)\n",
    "                    tree_NMPs.obs.loc['TLS_TLSCL_' + child.name, 'Somite_Neural_ratio'] = (somitic_count + 1) / (neural_count + 1)\n",
    "                        \n",
    "\n",
    "                    if states == set(['Neural', 'NMPs']):\n",
    "                        tree_NMPs.obs.loc['TLS_TLSCL_' + child.name, 'NMP Type'] = 'NMPs-Neural'\n",
    "                    elif states == set(['Somitic', 'NMPs']):\n",
    "                        tree_NMPs.obs.loc['TLS_TLSCL_' + child.name, 'NMP Type'] = 'NMPs-Somitic'\n",
    "                    elif states == set(['Neural', 'Somitic', 'NMPs']):\n",
    "                        tree_NMPs.obs.loc['TLS_TLSCL_' + child.name, 'NMP Type'] = 'NMPs-Bipotent'\n",
    "                    elif states == set(['NMPs']):\n",
    "                        tree_NMPs.obs.loc['TLS_TLSCL_' + child.name, 'NMP Type'] = 'NMPs-Renewing'\n",
    "                    else:\n",
    "                        tree_NMPs.obs.loc['TLS_TLSCL_' + child.name, 'NMP Type'] = 'Dropped'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fbac9bf5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/seaborn/distributions.py:316: UserWarning: Dataset has 0 variance; skipping density estimate. Pass `warn_singular=False` to disable this warning.\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "# NMP specific density plots after filtering the NMPs that share siblings that contain endoderm, endothelial, or PGC\n",
    "# classifying these NMPs using only sibling cells (not taking into account sibling nodes)\n",
    "fig, ax = plt.subplots(2, 2, figsize = (20, 10))\n",
    "order = ['NMPs-Bipotent', 'NMPs-Somitic', 'NMPs-Neural', 'NMPs-Renewing']\n",
    "fig.suptitle('Classifying NMPs using only sibling cells') # or plt.suptitle('Main title')\n",
    "\n",
    "ax1 = ax[0, 0]\n",
    "temp = tree_NMPs.obs[tree_NMPs.obs['orig.ident'] == 'TLS_120h_rep1']\n",
    "labels = []\n",
    "for i in order:\n",
    "    sns.kdeplot(temp[(temp['NMP Type'] == i)]['NT-Somite_Traj'], ax = ax1)\n",
    "    labels.append(i + ' n = ' + str(len(temp[(temp['NMP Type'] == i)]['NT-Somite_Traj'])))\n",
    "ax1.legend(labels)\n",
    "ax1.set_title('TLS1')\n",
    "ax1.set_xlim(-10, 8)\n",
    "ax1.set_ylim(0, 0.5)\n",
    "\n",
    "ax2 = ax[0, 1]\n",
    "temp = tree_NMPs.obs[tree_NMPs.obs['orig.ident'] == 'TLS_120h_rep2']\n",
    "labels = []\n",
    "for i in order:\n",
    "    if len(temp[(temp['NMP Type'] == i)]['NT-Somite_Traj']) > 0:\n",
    "        sns.kdeplot(temp[(temp['NMP Type'] == i)]['NT-Somite_Traj'], ax = ax2)\n",
    "        labels.append(i + ' n = ' + str(len(temp[(temp['NMP Type'] == i)]['NT-Somite_Traj'])))\n",
    "ax2.legend(labels)\n",
    "ax2.set_title('TLS2')\n",
    "ax2.set_xlim(-10, 8)\n",
    "ax2.set_ylim(0, 0.5)\n",
    "\n",
    "ax3 = ax[1, 0]\n",
    "temp = tree_NMPs.obs[(tree_NMPs.obs['orig.ident'] == 'TLS_TLSCL') & (tree_NMPs.obs['TLSCL'] == 'TLS')]\n",
    "labels = []\n",
    "for i in order:\n",
    "    sns.kdeplot(temp[(temp['NMP Type'] == i)]['NT-Somite_Traj'], ax = ax3)\n",
    "    labels.append(i + ' n = ' + str(len(temp[(temp['NMP Type'] == i)]['NT-Somite_Traj'])))\n",
    "ax3.legend(labels)\n",
    "ax3.set_title('TLS MultiSeq')\n",
    "ax3.set_xlim(-10, 8)\n",
    "ax3.set_ylim(0, 0.5)\n",
    "\n",
    "ax4 = ax[1, 1]\n",
    "temp = tree_NMPs.obs[(tree_NMPs.obs['orig.ident'] == 'TLS_TLSCL') & (tree_NMPs.obs['TLSCL'] == 'TLSCL')]\n",
    "labels = []\n",
    "for i in order:\n",
    "    sns.kdeplot(temp[(temp['NMP Type'] == i)]['NT-Somite_Traj'], ax = ax4)\n",
    "    labels.append(i + ' n = ' + str(len(temp[(temp['NMP Type'] == i)]['NT-Somite_Traj'])))\n",
    "ax4.legend(labels)\n",
    "ax4.set_title('TLSCL MultiSeq')\n",
    "ax4.set_xlim(-10, 8)\n",
    "ax4.set_ylim(0, 0.5)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/2_pseudotime/pseudotime_NMP_type_density.pdf', dpi = 300)\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77f435a6",
   "metadata": {},
   "source": [
    "# Plot NMP types over the pseudotime umap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d607bc4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot each of the dataset's NMPs over th epseudotime\n",
    "temp = tree_NMPs.copy()\n",
    "\n",
    "fig, axs = plt.subplots(1, 4, figsize=(20, 5))\n",
    "\n",
    "count = 0\n",
    "done = False\n",
    "for ident in ['TLS_120h_rep1', 'TLS_120h_rep2', 'TLS_TLSCL', 'TLS_TLSCL']:\n",
    "    ax = axs[count]\n",
    "    ax.plot(TLS_integrated.obsm['X_umap'][:,0], TLS_integrated.obsm['X_umap'][:,1], '.', color= \"lightgray\", alpha = 0.8, ms = 5, rasterized = True, markeredgewidth = 0.0)\n",
    "    \n",
    "    if ident == 'TLS_TLSCL':\n",
    "        if done:\n",
    "            temp_ident = temp[(temp.obs['orig.ident'] == ident) & (temp.obs['TLSCL'] == 'TLSCL')].copy()\n",
    "        else:\n",
    "            temp_ident = temp[(temp.obs['orig.ident'] == ident) & (temp.obs['TLSCL'] == 'TLS')].copy()\n",
    "            done = True\n",
    "    else:\n",
    "        temp_ident = temp[(temp.obs['orig.ident'] == ident)].copy()\n",
    "    \n",
    "    ax.plot(temp_ident.obsm['X_umap'][:,0], temp_ident.obsm['X_umap'][:,1], '.', alpha = 1, ms = 5, markeredgewidth = 0.0)\n",
    "    ax.set_title(ident + ' NMPs')\n",
    "    ax.axis([-15, 11, -15, 15])\n",
    "    ax.set_yticks([])\n",
    "    ax.set_xticks([])\n",
    "    count += 1\n",
    "\n",
    "plt.savefig('/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/2_pseudotime/umap_NMP_types.pdf', dpi = 300)\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "48e4def0",
   "metadata": {},
   "outputs": [],
   "source": [
    "colors = {'NMPs-Renewing': 'turquoise',\n",
    "          'NMPs-Neural': 'green',\n",
    "          'NMPs-Somitic': 'magenta',\n",
    "          'NMPs-Bipotent': 'blue',\n",
    "          'Dropped': 'gray'}\n",
    "\n",
    "temp = tree_NMPs[(tree_NMPs.obs['orig.ident'] == 'TLS_TLSCL') & (tree_NMPs.obs['TLSCL'] == 'TLS')].copy()\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 5))\n",
    "\n",
    "ax.plot(TLS_integrated.obsm['X_umap'][:,0], TLS_integrated.obsm['X_umap'][:,1], '.', color= \"lightgray\", alpha = 0.8, ms = 5, rasterized = True, markeredgewidth = 0.0)\n",
    "    \n",
    "for NMP in ['Dropped', 'NMPs-Renewing', 'NMPs-Neural', 'NMPs-Bipotent', 'NMPs-Somitic']:\n",
    "    NMP_temp = temp[temp.obs['NMP Type'] == NMP].copy()\n",
    "    ax.plot(NMP_temp.obsm['X_umap'][:,0], NMP_temp.obsm['X_umap'][:,1], '.', alpha = 1, color = colors[NMP], ms = 7.5, markeredgecolor = 'black', markeredgewidth = 0.25, label = NMP)\n",
    "\n",
    "ax.set_title('NMP Types')\n",
    "ax.legend()\n",
    "ax.axis([-15, 11, -15, 15])\n",
    "ax.set_yticks([])\n",
    "ax.set_xticks([])\n",
    "\n",
    "plt.savefig('/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/2_pseudotime/umap_combined_TLS-M_NMP_types.pdf', dpi = 300)\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9937baeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = tree_NMPs[(tree_NMPs.obs['orig.ident'] == 'TLS_TLSCL') & (tree_NMPs.obs['TLSCL'] == 'TLS')].copy()\n",
    "\n",
    "fig, ax = plt.subplots(1, 1, figsize=(6, 5))\n",
    "\n",
    "ax.scatter(TLS_integrated.obsm['X_umap'][:,0], TLS_integrated.obsm['X_umap'][:,1], color= \"lightgray\", alpha = 0.5, s = 5)\n",
    "\n",
    "x = TLS_integrated[(TLS_integrated.obs['cell_state'] != 'Endoderm') & (TLS_integrated.obs['cell_state'] != 'Unknown') & (TLS_integrated.obs['cell_state'] != 'Endothelial') & (TLS_integrated.obs['cell_state'] != 'PCGLCs')].obsm['X_umap'][:,0]\n",
    "y =  TLS_integrated[(TLS_integrated.obs['cell_state'] != 'Endoderm') & (TLS_integrated.obs['cell_state'] != 'Unknown') & (TLS_integrated.obs['cell_state'] != 'Endothelial') & (TLS_integrated.obs['cell_state'] != 'PCGLCs')].obsm['X_umap'][:,1]\n",
    "z = TLS_integrated[(TLS_integrated.obs['cell_state'] != 'Endoderm') & (TLS_integrated.obs['cell_state'] != 'Unknown') & (TLS_integrated.obs['cell_state'] != 'Endothelial') & (TLS_integrated.obs['cell_state'] != 'PCGLCs')].obs['NT-Somite_Traj']\n",
    "points = ax.scatter(x, y, c = z, cmap = 'viridis', alpha = 0.5, s = 5)\n",
    "\n",
    "for NMP in ['Dropped', 'NMPs-Renewing', 'NMPs-Neural', 'NMPs-Bipotent', 'NMPs-Somitic']:\n",
    "    NMP_temp = temp[temp.obs['NMP Type'] == NMP].copy()\n",
    "    ax.plot(NMP_temp.obsm['X_umap'][:,0], NMP_temp.obsm['X_umap'][:,1], '.', color = colors[NMP], alpha = 0.8, markersize = 7.5, markeredgecolor = 'black', markeredgewidth = 0.5, label = NMP)\n",
    "\n",
    "ax.set_title('NMPs on a Tree')\n",
    "ax.legend()\n",
    "ax.axis([-15, 11, -15, 15])\n",
    "ax.set_yticks([])\n",
    "ax.set_xticks([])\n",
    "fig.colorbar(points)\n",
    "\n",
    "plt.savefig('/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/2_pseudotime/umap_combined_TLS-M_NMP_types_over_pseudotime.pdf', dpi = 300)\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2012ff1c",
   "metadata": {},
   "source": [
    "# Investigating monocle modules across the pseudotime\n",
    "\n",
    "- From monocle, I have created a dataframe of gene modules that vary significantly across the pseudotime. This table can be found here:\n",
    "\n",
    "'/Genomics/chanlab/blaw/TLS/raw_data/monocle_results/TLS_TLSCL_1_2_Time_Integrated_monocle_full_subset_gene_modules_high_res_genes_above_0.1morans.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2f9a5cf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "module_genes_df = pd.read_csv('/Genomics/chanlab/blaw/TLS/raw_data/monocle_results/TLS_TLSCL_1_2_Time_Integrated_monocle_full_subset_gene_modules_high_res_genes_above_0.1morans.txt', index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d5337f67",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/scanpy/tools/_score_genes.py:151: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  for cut in np.unique(obs_cut.loc[gene_list]):\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n"
     ]
    }
   ],
   "source": [
    "modules = module_genes_df['module'].unique()\n",
    "\n",
    "# use the scanpy score genes function to calculate the module score of each module for each cell\n",
    "for temp_module in modules:\n",
    "    gene_ids = module_genes_df[module_genes_df['module'] == temp_module]['id']\n",
    "    sc.tl.score_genes(NMP_states, gene_ids, ctrl_size = len(gene_ids), score_name = 'module_{}_score'.format(temp_module), use_raw = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "23572dba",
   "metadata": {},
   "outputs": [],
   "source": [
    "with rc_context({'figure.figsize': (5, 5)}):\n",
    "    sc.pl.umap(NMP_states, color=['module_1_score', 'module_2_score', 'module_3_score', 'module_4_score', 'module_5_score',\n",
    "                                  'module_6_score', 'module_7_score', 'module_8_score', 'module_9_score', 'module_10_score',\n",
    "                                  'module_11_score', 'module_12_score', 'module_13_score', 'module_14_score', 'module_15_score',\n",
    "                                  'module_16_score', 'module_17_score', 'module_18_score', 'module_19_score', 'module_20_score',\n",
    "                                  'module_21_score', 'module_22_score', 'module_23_score', 'module_24_score', 'module_25_score',\n",
    "                                  'module_26_score', 'module_27_score', 'module_28_score'], save = '_module_scores', show = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "089fa6d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n"
     ]
    }
   ],
   "source": [
    "# investigate module scores vs NMP types in TLS M\n",
    "temp = NMP_states[(NMP_states.obs['orig.ident'] == 'TLS_TLSCL') & (NMP_states.obs['TLSCL'] == 'TLS')].copy()\n",
    "temp.obs['NMP Type'] = 'Non-Tree'\n",
    "\n",
    "for i in tree_NMPs.obs.index:\n",
    "    if i in temp.obs.index:\n",
    "        temp.obs.loc[i, 'NMP Type'] = tree_NMPs.obs.loc[i, 'NMP Type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "19ad5888",
   "metadata": {},
   "outputs": [],
   "source": [
    "for module in [14, 7, 1, 4]:\n",
    "    data = []\n",
    "\n",
    "    cell_states = ['aPSM', 'pPSM', 'NMPs-Somitic', 'NMPs-Neural', 'NeuralTube1', 'NeuralTube2']\n",
    "    for i in cell_states:\n",
    "        if i.startswith('NMPs'):\n",
    "            data.append(temp.obs[temp.obs['NMP Type'] == i]['module_{}_score'.format(module)])\n",
    "        else:\n",
    "            data.append(temp.obs[temp.obs['cell_state'] == i]['module_{}_score'.format(module)])\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    plt.boxplot(data, labels = cell_states)\n",
    "    plt.title('Module {} Score - TLS M'.format(module))\n",
    "    plt.xlabel('Module Score')\n",
    "    plt.tight_layout()\n",
    "    plt.savefig('/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/2_pseudotime/module_{}_boxplots_NMP_Types_with_bookends.pdf'.format(module), dpi = 300)\n",
    "    #plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "5fdd956a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot modules for just NMP types\n",
    "for module in [14, 7, 1, 4]:\n",
    "    data = []\n",
    "\n",
    "    cell_states = ['NMPs-Renewing', 'NMPs-Somitic', 'NMPs-Neural', 'NMPs-Bipotent']\n",
    "    for i in cell_states:\n",
    "        if i.startswith('NMPs'):\n",
    "            data.append(temp.obs[temp.obs['NMP Type'] == i]['module_{}_score'.format(module)])\n",
    "        else:\n",
    "            data.append(temp.obs[temp.obs['cell_state'] == i]['module_{}_score'.format(module)])\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    plt.boxplot(data, labels = cell_states)\n",
    "    plt.title('Module {} Score - TLS M'.format(module))\n",
    "    plt.xlabel('Module Score')\n",
    "    plt.tight_layout()\n",
    "    plt.savefig('/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/2_pseudotime/module_{}_boxplots_NMP_Types.pdf'.format(module), dpi = 300)\n",
    "    #plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51d6ca16",
   "metadata": {},
   "source": [
    "# Correlate the module score with the ratio of somite / neural"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b0ae93b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/python/3.9.11-mmchan/lib/python3.9/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n"
     ]
    }
   ],
   "source": [
    "# Look at only TLSM cells that are on the tree\n",
    "test_tree_cells = []\n",
    "\n",
    "for i in temp.obs.index:\n",
    "    if i in tree_NMPs.obs.index:\n",
    "        test_tree_cells.append(i)\n",
    "        \n",
    "temp = temp[temp.obs.index.isin(test_tree_cells)].copy()\n",
    "\n",
    "temp.obs['frac_Neural'] = ''\n",
    "temp.obs['frac_Somitic'] = ''\n",
    "temp.obs['Somite_Neural_ratio'] = ''\n",
    "\n",
    "for i in temp.obs.index:\n",
    "    temp.obs.loc[i, 'frac_Neural'] = tree_NMPs.obs.loc[i, 'frac_Neural']\n",
    "    temp.obs.loc[i, 'frac_Somitic'] = tree_NMPs.obs.loc[i, 'frac_Somitic']\n",
    "    temp.obs.loc[i, 'Somite_Neural_ratio'] = tree_NMPs.obs.loc[i, 'Somite_Neural_ratio']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f49380b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test the correlation between the somite neural ratio and module score in TLS M tree NMPs\n",
    "module_corr = pd.DataFrame(index = modules, columns = ['slope', 'statistic', 'pval', 'adj_pval'])\n",
    "\n",
    "for temp_module in module_corr.index:\n",
    "    slope, intercept, r, p, se = stats.linregress(temp.obs['module_{}_score'.format(temp_module)].tolist(), temp.obs['Somite_Neural_ratio'].tolist())\n",
    "\n",
    "    module_corr.loc[temp_module, 'statistic'] = r\n",
    "    module_corr.loc[temp_module, 'pval'] = p\n",
    "    module_corr.loc[temp_module, 'slope'] = slope\n",
    "    \n",
    "# adjust the pval using multiple hypothesis testing\n",
    "module_corr['adj_pval'] = module_corr['pval'] * len(module_corr.index)\n",
    "\n",
    "module_corr.to_csv('/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/2_pseudotime/TreeNMPs_TLSM_module_correlations_full_pseudotime_high_res_genes_above_0.1morans.txt', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "7f4d1773",
   "metadata": {},
   "outputs": [],
   "source": [
    "bars = []\n",
    "heights = []\n",
    "for bar in module_corr.sort_values(by = 'slope').index:\n",
    "    heights.append(module_corr.loc[bar, 'slope'])\n",
    "    bars.append(str(bar))\n",
    "\n",
    "fig, ax = plt.subplots(figsize = (10, 5))\n",
    "plt.bar(x = bars, height = heights)\n",
    "plt.xlabel('module number')\n",
    "plt.ylabel('Slope')\n",
    "plt.title('Module Correlation Slopes')\n",
    "plt.savefig('/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/2_pseudotime/module_slope_full_pseudotime_high_res_genes_above_0.1morans.pdf', dpi = 300)\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5244a618",
   "metadata": {},
   "source": [
    "# Investigating gene expression of NMPs over the pseudotime\n",
    "\n",
    "- In this section I will extract the average expression of NMPs across the pseudotime for genes identified in the module analysis\n",
    "\n",
    "- I will focus on the -10 to 8range in the pseudotime, removing non-NMPs from the dataset\n",
    "\n",
    "- I will average the expression of all cells in 1 step bins over the pseudotime\n",
    "\n",
    "- Remove the 96h dataset and remove the TLSCL dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c7f7142f",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mature_NMPs = NMPs[(NMPs.obs['orig.ident'].isin(['mGast_120h', 'mGast_108h', 'TLS_120h_rep1', 'TLS_120h_rep2', 'TLS_TLSCL']))\n",
    "                   & (NMPs.obs['NT-Somite_Traj'] >= -10) & (NMPs.obs['NT-Somite_Traj'] <= 8) & (NMPs.obs['TLSCL'] != 'TLSCL')\n",
    "                   & (NMPs.obs['MultiSeqGroup'] != 'Doublet') & (NMPs.obs['MultiSeqGroup'] != 'Negative')].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2a6a70a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "ordered_cells = list(Mature_NMPs.obs.sort_values(by = 'NT-Somite_Traj').index)\n",
    "\n",
    "# split the cells into 1 step bins along the pseudotime\n",
    "itr_dict = {}\n",
    "\n",
    "previous = 0\n",
    "for i in range(1, 19):\n",
    "    lower = -10 + previous\n",
    "    upper = -10 + i\n",
    "    cells = list(Mature_NMPs.obs[(Mature_NMPs.obs['NT-Somite_Traj'] >= lower) & (Mature_NMPs.obs['NT-Somite_Traj'] < upper)].index)\n",
    "    previous = i\n",
    "    temp = str(lower) + '_' + str(upper)\n",
    "    itr_dict[temp] = cells\n",
    "\n",
    "with open('/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/2_pseudotime/NMPs_pseudotime_-10-8_1_step_bins.pickle', 'wb') as handle:\n",
    "    pickle.dump(itr_dict, handle, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "7f1491b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the avg expression of each of the 1 step bins\n",
    "avg_itr_expr_df = pd.DataFrame(columns = Mature_NMPs.var['features'], index = itr_dict.keys())\n",
    "\n",
    "for i in avg_itr_expr_df.index:\n",
    "    cells = itr_dict[i]\n",
    "    temp = Mature_NMPs[Mature_NMPs.obs.index.isin(cells)].copy()\n",
    "    \n",
    "    expr = temp.X.todense()\n",
    "\n",
    "    avg_itr_expr_df.loc[i] = expr.mean(axis = 0)\n",
    "\n",
    "avg_itr_expr_df.to_csv('/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/2_pseudotime/NMPs_pseudotime_-10-8_1_step_avg_expr.txt', sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0faa5dc",
   "metadata": {},
   "source": [
    "# Plot the expression of these bins for NMPS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d49d59c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "avg_itr_expr_df = pd.read_csv('/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/2_pseudotime/NMPs_pseudotime_-10-8_1_step_avg_expr.txt', sep = '\\t', index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e4f42323",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate the zscores of genes\n",
    "z_itr_expr_df = avg_itr_expr_df.copy()\n",
    "\n",
    "mean_vals = z_itr_expr_df.mean(axis = 0)\n",
    "std_vals = z_itr_expr_df.std(axis = 0)\n",
    "\n",
    "z_itr_expr_df = (z_itr_expr_df - mean_vals) / (std_vals)\n",
    "z_itr_expr_df.fillna(0, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "09e8405b",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = z_itr_expr_df.copy()\n",
    "genes_14 = module_genes_df[module_genes_df['module'] == 14]['id'].tolist()\n",
    "genes_4 = module_genes_df[module_genes_df['module'] == 4]['id'].tolist()\n",
    "#genes_1 = module_genes_df[module_genes_df['module'] == 1]['id'].tolist()\n",
    "genes_7 = module_genes_df[module_genes_df['module'] == 7]['id'].tolist()\n",
    "genes_of_interest = genes_14 + genes_4 + genes_7\n",
    "temp = temp[genes_of_interest]\n",
    "#temp.drop(index = ['-15_-14', '-14_-13', '-13_-12', '-12_-11', '-11_-10', '8_9', '9_10', '10_11', '11_12', '12_13', '13_14', '14_15'], inplace = True)\n",
    "\n",
    "vcenter = 0\n",
    "#vmin, vmax = temp.min().min(), temp.max().max()\n",
    "vmin, vmax = -2, 3\n",
    "normalize = mcolors.TwoSlopeNorm(vcenter=vcenter, vmin=vmin, vmax=vmax)\n",
    "colormap = cm.RdYlBu_r\n",
    "\n",
    "g = sns.clustermap(temp.T, figsize = (10, 60), col_cluster = False, row_cluster = False, norm=normalize, cmap=colormap)\n",
    "plt.subplots_adjust(bottom = 0.3, right = 0.7)\n",
    "plt.savefig('/Genomics/chanlab/blaw/TLS/data/integrated_scRNA/2_pseudotime/module_14-4-7_NMPs_pseudotime_high_res_genes_above_0.1morans.pdf', dpi = 300)\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
